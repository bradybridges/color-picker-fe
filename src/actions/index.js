export const addProjects = (projects) => ({
  type: 'ADD_PROJECTS',
  projects,
});

export const addPalettes = (palettes) => ({
  type: 'ADD_PALETTES',
  palettes,
});
