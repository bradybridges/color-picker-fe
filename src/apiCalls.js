export const getProjects = async () => {
  const response = await fetch('https://color-picker-backend.herokuapp.com/api/v1/projects');
  if(!response.ok) {
    return Error('Failed to fetch projects');
  }
  const projects = await response.json();
  return projects;
};

export const getPalettes = async () => {
  const response = await fetch('https://color-picker-backend.herokuapp.com/api/v1/palettes');
  if(!response.ok) {
    return Error('Failed to fetch palettes');
  }
  const palettes = await response.json();
  return palettes;
};

export const getProject = async (id) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/projects/${id}`);
  if(!response.ok) {
    return Error('Failed to fetch pallete');
  }
  const palette = await response.json();
  return palette;
};

export const getPalette = async (paletteId) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/palettes/${paletteId}`);
  if (!response.ok) {
    return Error('Failed to fetch palette');
  }
  const palette = await response.json();
  return palette;
};

export const postProject = async (name) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/projects`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  });
  if (!response.ok) {
    return Error('Failed to post new project');
  }
  const message = await response.json();
  console.log('message', message)
  return message;
};

export const postPalette = async (project_id, palette_name, color_1, color_2, color_3, color_4, color_5) => {
  const response = await fetch('https://color-picker-backend.herokuapp.com/api/v1/palettes', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(
      {
        project_id,
        palette_name,
        color_1,
        color_2,
        color_3,
        color_4,
        color_5,
      },
    ),
  });
  if (!response.ok) {
    return response;
  }
  const message = await response.json();
  return message;
};

export const deleteProject = async (id) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/projects/${id}`, {
    method: 'DELETE',
  });
  if (!response.ok) {
    return Error('Failed to delete project');
  }
  const message = await response.json();
  return message;
};

export const deletePalette = async (id) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/palettes/${id}`, {
    method: 'DELETE',
  });
  if (!response.ok) {
    return Error('Failed to delete palette');
  }
  const message = await response.json();
  return message;
};

export const patchProject = async () => {

};

export const patchPalette = async () => {

};
