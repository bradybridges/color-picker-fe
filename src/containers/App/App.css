.main-content {
  background-color: black;
  overflow: scroll;
  height: 100vh;
}

.button-container {
  margin: auto;
  text-align: center;
}

.generate-btn {
  margin: 0 auto;
  background: black;
  color: #2d9fcc;
  font-size: 1.5em;
  border-radius: 3px 3px 3px 3px;
  font-family: 'Permanent Marker';
  border: 1px solid grey;
}

.main-header {
  color: #2d9fcc;
  font-family: 'Rock Salt', cursive;
  /* font-family: 'Permanent Marker', cursive; */
  font-size: 3em;
  margin: 10px 40px;
}

.user-projects {
  display: flex;
  flex-direction: column;
  /* grid-gap: 20px 15px; */
  /* grid-template-columns: repeat(auto-fit, minmax(120px, 0.3fr)); */
}