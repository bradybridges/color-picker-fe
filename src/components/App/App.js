import React, { Component } from 'react';
import './App.css';
import { getProjects, getPalettes } from '../../apiCalls';
import * as actions from '../../actions/index';
import { connect } from 'react-redux';

export class App extends Component {
 
  componentDidMount = async () => {
    const projects = await getProjects();
    this.props.setProjects(projects);
    const palettes = await getPalettes();
    this.props.setPalettes(palettes);
  }

  render() {
    return (
      <h1>Hello</h1>
    );
  }
}

export const mapState = (state) => ({
  projects: state.projects,
  palettes: state.palettes,
});

export const mapDispatch = (dispatch) => ({
  setProjects: (projects) => dispatch(actions.addProjects(projects)),
  setPalettes: (palettes) => dispatch(actions.addPalettes(palettes)),
});

export default connect(mapState, mapDispatch)(App);
