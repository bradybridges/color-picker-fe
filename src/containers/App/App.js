import React, { Component } from 'react';
import './App.css';
import * as api from '../../apiCalls';
import * as actions from '../../actions/index';
import { connect } from 'react-redux';
import PaletteContainer from '../PaletteContainer/PaletteContainer';
import SavePaletteForm from '../SavePaletteForm/SavePaletteForm';
import ProjectContainer from '../ProjectContainer/ProjectContainer';
import SaveProjectForm from '../SaveProjectForm/SaveProjectForm'

export class App extends Component {
  constructor() {
    super();
    this.state = {
      isLoading: true,
      tempColors: [
        {name: 'color_1', color: '#000000', isLocked: false },
        {name: 'color_2', color: '#000000', isLocked: false },
        {name: 'color_3', color: '#000000', isLocked: false },
        {name: 'color_4', color: '#000000', isLocked: false },
        {name: 'color_5', color: '#000000', isLocked: false },
      ],
    };
  }
 
  componentDidMount = async () => {
    const projects = await api.getProjects();
    const palettes = await api.getPalettes();
    this.props.setProjects(projects);
    this.props.setPalettes(palettes);
    this.setState({ isLoading: false });
  }

  generatePalette = () => {
    this.props.tempPalette.forEach((color, i) => {
      if(!color.isLocked) {
        const randomColor = this.generateColor();
        //this is so we don't modify store directly but call actions to handle this
        let updatedPalette = this.props.tempPalette.map((color) => color);
        updatedPalette[i].color = randomColor;
        this.props.setTempPalette(updatedPalette);
      }
    });
  }

  generateColor = () => {
    const options = ['A','B','C','D','E','F','0','1','2','3','4','5','6','7','8','9'];
    let newColor = [];
    for(let i = 0; i < 6; i++) { 
      const randomIndex = Math.floor(Math.random() * options.length);
      newColor.push(options[randomIndex]);
    }
    newColor.unshift('#');
    return newColor.join('');
  }

  render() {
    return (
      <main className='main-content'>
        {this.state.isLoading && <h1>Loading...</h1>}
        <h1 className='main-header'>Color Coordinated</h1>
        <div className='button-container'>
          <button onClick={this.generatePalette} className='generate-btn'>Generate</button>
        </div>
        <PaletteContainer />
        <SaveProjectForm postProject={api.postProject}/>
        <SavePaletteForm savePalette={api.postPalette}/>
        <section className='user-projects'>
          <ProjectContainer />
        </section>
      </main>
    );
  }
}

export const mapState = (state) => ({
  projects: state.projects,
  palettes: state.palettes,
  tempPalette: state.tempPalette,
});

export const mapDispatch = (dispatch) => ({
  setProjects: (projects) => dispatch(actions.addProjects(projects)),
  setPalettes: (palettes) => dispatch(actions.addPalettes(palettes)),
  setTempPalette: (palette) => dispatch(actions.setTempPalette(palette)),
});

export default connect(mapState, mapDispatch)(App);
