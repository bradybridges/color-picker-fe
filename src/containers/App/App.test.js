import React from 'react';
import ReactDOM from 'react-dom';
import {App, mapState, mapDispatch } from './App';
import { shallow } from 'enzyme'
import { addProjects, addPalettes, setTempPalette } from '../../actions'
import * as api from '../../apiCalls'
//getProjects, getPalettes


//props: projects, palettes, tempPalette, setProjects, setPalettes
//setTempPalette, getProjects, getPalettes

jest.mock('../../apiCalls')

describe('App', () => {
  describe('App component', () => {
    let wrapper;

    beforeEach(() => {
       api.getProjects.mockImplementation(() => {
        return Promise.resolve([
          {
            id: 1,
            name: 'a project'
          }, 
          {
            id: 2,
            name: 'another project'
          }
        ])
      })

       api.getPalettes.mockImplementation(() => {
        return Promise.resolve([
          {
            id: 1,
            project_id: 1,
            name: 'a palette',
            color_1: '#FFFFFF',
            color_2: '#111111',
            color_3: '#222222',
            color_4: '#333333',
            color_5: '#444444',
          },
          {
            id: 2,
            project_id: 1,
            name: 'some other palette',
            color_1: '#000000',
            color_2: '#AAAAAA',
            color_3: '#BBBBBB',
            color_4: '#CCCCCC',
            color_5: '#DDDDDD',
          }
        ])
      })

      const mockProjects = [
        {
          id: 1,
          name: 'a project'
        }, 
        {
          id: 2,
          name: 'another project'
        }
      ]

      const mockPalettes = [
        {
          id: 1,
          project_id: 1,
          name: 'a palette',
          color_1: '#FFFFFF',
          color_2: '#111111',
          color_3: '#222222',
          color_4: '#333333',
          color_5: '#444444',
        },
        {
          id: 2,
          project_id: 1,
          name: 'some other palette',
          color_1: '#000000',
          color_2: '#AAAAAA',
          color_3: '#BBBBBB',
          color_4: '#CCCCCC',
          color_5: '#DDDDDD',
        }
      ]

      const mockTempPalette = [
        {name: 'color_1', color: '#AAAAAA', isLocked: false},
        {name: 'color_2', color: '#AAAAAA', isLocked: false},
        {name: 'color_3', color: '#AAAAAA', isLocked: false},
        {name: 'color_4', color: '#AAAAAA', isLocked: false},
        {name: 'color_5', color: '#AAAAAA', isLocked: false},
      ]

      const mockSetProjects = jest.fn()
      const mockSetPalettes = jest.fn()
      const mockSetTempPalette = jest.fn()

      wrapper = shallow(<App 
        projects={mockProjects} 
        palettes={mockPalettes} 
        tempPalette={mockTempPalette}
        setProjects={mockSetProjects}
        setPalettes={mockSetPalettes}
        setTempPalette={mockSetTempPalette} 
        />)
    })

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should fetch projects after mounting', () => {
      expect(api.getProjects).toHaveBeenCalled()
    })

    it('should fetch palettes after mounting', () => {
      expect(api.getPalettes).toHaveBeenCalled()
    })

    it('should call setProjects after mounting', () => {
      expect(wrapper.instance().props.setProjects).toHaveBeenCalled()
      expect(wrapper.instance().props.setProjects).toHaveBeenCalledWith([
        {
          id: 1,
          name: 'a project'
        }, 
        {
          id: 2,
          name: 'another project'
        }
      ])
    })

    it('should call setPalettes after mounting', () => {
      expect(wrapper.instance().props.setPalettes).toHaveBeenCalled()
      expect(wrapper.instance().props.setPalettes).toHaveBeenCalledWith([
        {
          id: 1,
          project_id: 1,
          name: 'a palette',
          color_1: '#FFFFFF',
          color_2: '#111111',
          color_3: '#222222',
          color_4: '#333333',
          color_5: '#444444',
        },
        {
          id: 2,
          project_id: 1,
          name: 'some other palette',
          color_1: '#000000',
          color_2: '#AAAAAA',
          color_3: '#BBBBBB',
          color_4: '#CCCCCC',
          color_5: '#DDDDDD',
        }
      ])
    })

    it('should call the generatePalette method when the button is clicked', () => {
      wrapper.instance().generatePalette = jest.fn()

      wrapper.instance().forceUpdate()

      wrapper.find('button').simulate('click')

      expect(wrapper.instance().generatePalette).toHaveBeenCalled()
    })
    
    it('should call setTempPalette when the button is clicked and generatePalette is called', () => {
      wrapper.instance().generatePalette()

      expect(wrapper.instance().props.setTempPalette).toHaveBeenCalled()
    })
  })
})


