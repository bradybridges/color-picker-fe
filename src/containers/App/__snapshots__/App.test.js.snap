// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App component should match the snapshot 1`] = `
<main
  className="main-content"
>
  <h1
    className="main-header"
  >
    Color Coordinated
  </h1>
  <div
    className="button-container"
  >
    <button
      className="generate-btn"
      onClick={[Function]}
    >
      Generate
    </button>
  </div>
  <Connect(PaletteContainer) />
  <Connect(SaveProjectForm)
    postProject={[MockFunction]}
  />
  <Connect(SavePaletteForm)
    savePalette={[MockFunction]}
  />
  <section
    className="user-projects"
  >
    <Connect(ProjectContainer) />
  </section>
</main>
`;
